pkg/build/builder/dockerutil.go

```diff
@@ -62,15 +73,59 @@ func pushImage(client DockerClient, name string, authConfig docker.AuthConfigura
 	if glog.V(5) {
 		opts.OutputStream = os.Stderr
 	}
-	var err error
+
 	var retriableError = false
 
+	eventWatch, err := eventClient.Watch(
+		kapi.ListOptions{
+			FieldSelector: fields.ParseSelectorOrDie(
+				strings.Join([]string{
+					"involvedObject.kind==" + "ImageStream",
+					"involvedObject.namespace==" + ref.Namespace,
+					"involvedObject.name==" + ref.Name},
+					",")),
+		},
+	)
+	if err != nil {
+		glog.Errorf("Failed to watch for events: %v", err)
+	} else {
+		glog.V(5).Infof("Listening for events on image stream %s", ref.RepositoryName())
+		defer eventWatch.Stop()
+	}
+
 	for retries := 0; retries <= DefaultPushRetryCount; retries++ {
 		err = client.PushImage(opts, authConfig)
 		if err == nil {
 			return nil
 		}
 
+		waitEnd := time.Now().Add(time.Second * 2)
+
+		if eventWatch != nil && strings.Contains(err.Error(), "denied") {
+			glog.V(4).Infof("Push has been denied. Waiting for an event with proper error description.")
+			for time.Now().Before(waitEnd) {
+				select {
+				case e := <-eventWatch.ResultChan():
+					if e.Type != watchapi.Added {
+						continue
+					}
+					event, ok := e.Object.(*kapi.Event)
+					if !ok {
+						continue
+					}
+					if event.InvolvedObject.Kind != "ImageStream" || event.InvolvedObject.Namespace != ref.Namespace || event.InvolvedObject.Name != ref.Name {
+						continue
+					}
+
+					glog.V(4).Infof("Replacing error message %q with %q", err.Error(), event.Message)
+					err = errors.New(event.Message)
+					break
+				case <-time.After(waitEnd.Sub(time.Now())):
```

Comment:
It seems that the cases where the code will continue above are corner-cases (or maybe even impossible to happen) but still this isn't 100% correct. If you want the timeout channel to have full effect ie. work as expected independently of when other channels receive in a select, you need to define it outside the loop and receive from it here. (otherwise each time you receive from the other channels, this will retry to timeout from scratch).

https://play.golang.org/p/ZG5M1F27DC